### PR TITLE
test: Adds integration tests for Links API

### PR DESCRIPTION
**Pull request by wbreza**
Adds a new integration tests project to run integration tests for backend APIs.

- [x] Adds integration tests for Links API
- [x] Updates API pipeline to only run unit test project
- [x] Updates API pipeline to deploy to staging slot and auto swap after running integration tests.

### CODE REVIEW

File: api/README.md [hunk 1]
```diff
@@ -124,6 +123,39 @@ If everything was setup correctly, you should see a response that resembles the
...
+export INTTEST_AzureAdB2C__ClientId= 
+# The Azure B2C authority endpoint
+export INTTEST_AzureAdB2C__Authority= 
+# The scope to set on your access token
+export INTTEST_AzureAdB2C__Scope= 
+# The username of your test user account
+export INTTEST_AzureAdB2C__Username= 
+# The password of your test user account
+export INTTEST_AzureAdB2C__Password= 
+```
```
Comment (alkofahi) [inline api/README.md]: Some questions:
1) Out of curiosity, why double underscores before the attribute name?
2) Would it be possible to describe where they could get these values?
3) what is the test user account here?
4) I might be over-thinking it, but is there a better way than to pass the password here?

File: api/tests/LinkyLink.Integration.Tests/LinkApiTests.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+            var authResult = await app
+                .AcquireTokenByUsernamePassword(appScope, credentials.UserName, credentials.SecurePassword)
+                .ExecuteAsync();
+
+            return authResult.AccessToken;
+        }
+
+        [TestMethod]
+        public async Task Links_API_CRUD_Operations()
```
Comment (alkofahi) [inline api/tests/LinkyLink.Integration.Tests/LinkApiTests.cs]: Can we break this test into multiple methods  where each performs a certain part of the test and then call all of them in sequence within this test?

File: pipelines/api-pipelines.yml [hunk 3]
```diff
@@ -100,29 +100,57 @@ stages:
...
               downloadPath: '$(Build.ArtifactStagingDirectory)'
+
           - task: AzureRMWebAppDeployment@4
-            displayName: Deploy to App Service
+            displayName: Deploy to Staging Slot
             inputs:
               ConnectionType: 'AzureRM'
               azureSubscription: $(azureSubscription)
               appType: 'apiApp'
+              ResourceGroupName: 'rg-$(businessUnit)-$(appName)-$(environment)-$(region)'
```
Comment (alkofahi) [inline pipelines/api-pipelines.yml]: minor - would we be able to have a step to populate variables with the RG name, backend endpoint...etc and reuse those variables through out the pipeline?

File: api/tests/LinkyLink.Integration.Tests/LinkApiTests.cs (hunk 2 above)
Comment (wbreza) [reply api/tests/LinkyLink.Integration.Tests/LinkApiTests.cs]: Yes - I almost did the same :)

File: pipelines/api-pipelines.yml (hunk 3 above)
Comment (wbreza) [reply pipelines/api-pipelines.yml]: This runs inside a parallel matrix build so we couldn't create independently before that.

File: api/README.md (hunk 1 above)
Comment (wbreza) [reply api/README.md]: The double underscore is a convention used by the .net core configuration provider.
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-3.1#environment-variables-configuration-provider

I can add links to learn more about Azure B2C configuration.  I wouldn't want to duplicate existing documentation.

In order to test the API as an authenticated user we need to authenticate.  The integration tests use a OAuth password grant which allows us to use local user accounts associated to the Azure B2C tenant (not social accounts).

Without authenticating we also would not be able to test many of the endpoints for getting links by user, deleting links, etc.

Comment (wbreza) [reply api/README.md]: Updated readme with some more information.

File: api/tests/LinkyLink.Integration.Tests/LinkApiTests.cs (hunk 2 above)
Comment (alkofahi) [reply api/tests/LinkyLink.Integration.Tests/LinkApiTests.cs]: great minds think a like :sunglasses:

File: api/README.md (hunk 1 above)
Comment (alkofahi) [reply api/README.md]: Got it, makes sense and thanks for the response, insightful as always 😄

File: api/tests/LinkyLink.Integration.Tests/LinkApiTests.cs [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+            this.ValidateConfigurationItem(() => this.b2cConfig.Password, nameof(this.b2cConfig.Password));
+
+            this.client = new HttpClient
+            {
+                BaseAddress = new Uri(this.b2cConfig.BaseAddress)
+            };
+        }
+
+        [TestMethod]
+        public async Task Links_API_CRUD_Operations()
```
Comment (michaeldeongreen) [inline api/tests/LinkyLink.Integration.Tests/LinkApiTests.cs]: I wonder if it would be better if these were broken down into separate integration tests per endpoint?

Comment (wbreza) [reply api/tests/LinkyLink.Integration.Tests/LinkApiTests.cs]: The intention was to run them in a particular sequence so we can clean up after ourselves.

Comment (wbreza) [reply api/tests/LinkyLink.Integration.Tests/LinkApiTests.cs]: These integration tests / functional tests are essentially to run a smoke test against the API's to ensure they are working.  More intensive tests with positive/negative test cases should typically be done at the unit test level.
